### PR TITLE
[ENH] Add neuralforecast adapter and rnn forecaster

### PR DESCRIPTION
**Pull request by yarnabrina**
Adapter class to `neuralforecast` from `Nixtla` and concrete estimator for `RNN`.

Notes for reviewers (not in any particular order)

1. `neuralforecast` supports 3 types of exogenous features: static (constant over time), historical (unknown for prediction horizon) and future (known for prediction horizon). I felt the static type has no predictive capability if models are trained per series, and as far as I know, sktime will not be able to support historical type for framework compatibilities. Hence in this first draft of adaptation, I have added support for only future type.

2. `neuralforecast` has the capability of directly supporting panel/hierarchical datasets directly. I am not sure how to take benefit of that in sktime, so have not used it in this first draft. If it's easy enough to support, I can update if some references are shared, otherwise this can definitely be a extension.

3. ~`RNN` in `neuralforecast` supports specification of loss and validation loss, but they needs to be instances of [pytorch loss functions](https://nixtlaverse.nixtla.io/neuralforecast/losses.pytorch.html). I am not sure how set these defaults in instance, as soft dependencies can not be checked there. I have hence not exposed these parameters and rely on the defaults, but this is both important and common parameter. With some suggestions and references, I will like to enhance this in this draft itself.~ Added some support using internal attributes.

4. `neuralforecast` does support probablistic forecasts, but it seems that depends on choice of loss function (and may be something more, not sure). In this first draft, I have not added that support so far and this can be part of an extension.

5. `neuralforecast` has a public `NeuralForecast` class which accepts `pandas.DataFrame` objects of special schema directly, and has the ability to train multiple models and find best one. The underlying models do not support `pandas.DataFrame` directly and relies of `TimeSeriesDataset`, a sub-class of `torch.utils.data.Dataset`. For simplification with most other forecasters, I opted for the higher level class.

6. `neuralforecast` supports in-sample predictions, but warns that those may be inaccurate. Since the library itself is marked as `Alpha`, I have disabled entire support of in-sample predictions for now.

7. `neuralforecast` takes a lot of dependencies, and adding it in `forecasting` extra may cause too much installation for users who won't use it. But adding it in `dl` extra means new CI is not going to test it, so we need to decide what to do.

8. ~`neuralforecast` supports a lot of models, and (hopefully) all of them can be used with the adapter in this PR. However that is a lot of models, and it leads to the question of how should we organise them, by model or by dependency or by both or by something else.~ Modified only `dl` dependency similar to other recent PR's that affect torch/tensorflow etc.

### CODE REVIEW

**Comment by yarnabrina** [comment]
@AzulGarza @jmoralez if you can please take a look at this PR and share your reviews for this adaptation, especially if you have suggestions that will help in supporting a long term stable adapter, that will be much appreciated.

File: .github/workflows/test.yml
Comment (yarnabrina) [inline .github/workflows/test.yml:1]: Most of CI is disabled to quickly get test results for new addition. This is not part of the main PR goal, and will be reverted before merge. Ignore all changes in `.github`.

File: .github/workflows/test_module.yml
Comment (yarnabrina) [inline .github/workflows/test_module.yml:1]: Most of CI is disabled to quickly get test results for new addition. This is not part of the main PR goal, and will be reverted before merge. Ignore all changes in `.github`.

**Review by yarnabrina — COMMENTED** [review]
Most of CI is disabled to quickly get test results for new addition. This is not part of the main PR goal, and will be reverted before merge. Ignore all changes in `.github`.

NOTE: revert 2b756d996d612aa404b635c2162ce2c099a788b3 before marking ready for merge.

**Comment by yarnabrina** [comment]
FYI @fkiraly @benHeid @achieveordie : all tests pass for the relevant adapter and estimators in all environment combinations (CI is modified to ensure all tests for these run on all combinations of operating systems and python versions). Other than addressing reverting CI debug commit just before merge, I have no other new changes to add in this PR (other than review comments of course).

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks nice!

Few questions:

* the in-sample failures can be silenced by raising `NotImplementedError` if `not fh_is_all_in_sample()`. @benHeid had the same issue recently, probably we should do that in the base class.
* question: how would we test this in the intended end state?
* a bit unsettling that `freq` is an argument of `__init__`, rather than obtained from `X` or `fh`. Is the latter not possible?

File: sktime/forecasting/base/adapters/_neuralforecast.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+            guaranteed to have at least one column/variable
+            Exogeneous time series to fit to.
+
+        Returns
+        -------
+        self : _NeuralForecastAdapter
+            reference to self
+        """
+        if not fh.is_all_out_of_sample(cutoff=self.cutoff):
+            raise NotImplementedError("in-sample prediction is currently not supported")
```
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_neuralforecast.py:180]: I have this check in `_fit` already, and `fh` passed to `_predict` is unused. Cna you please tell what do you want me to change?

> the in-sample failures can be silenced by raising NotImplementedError if not fh_is_all_in_sample(). @benHeid had the same issue recently, probably we should do that in the base class.

I am confused a bit, why check for `if not fh_is_all_in_sample()`?

File: sktime/forecasting/neuralforecast.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+    >>>
+    >>> # getting point predictions
+    >>> model.predict(X=X_test)  # doctest: +SKIP
+    Predicting DataLoader 0: 100%|██████████████████████████████████| 1/1 [00:00<00:00, 198.64it/s]
+    1959    66241.984375
+    1960    66700.125000
+    1961    66550.195312
+    1962    67310.007812
+    Freq: A-DEC, Name: TOTEMP, dtype: float64
+    >>>
```
Comment (yarnabrina) [inline sktime/forecasting/neuralforecast.py:132]: This is how I visualised end users will use it. Is that what you are asking?

> question: how would we test this in the intended end state?

What tests do you suggest I should add?

File: sktime/forecasting/base/adapters/_neuralforecast.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+        )
+
+        algorithm_instance = self.algorithm_class(
+            fh,
+            alias=self.algorithm_name,
+            **self.algorithm_parameters,
+            **exogenous_parameters,
+        )
+
+        from neuralforecast import NeuralForecast
```
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_neuralforecast.py:143]: This takes frequency as init argument.

> a bit unsettling that freq is an argument of __init__, rather than obtained from X or fh. Is the latter not possible?

I won't say impossible, but with the amount of issues in `pandas` frequency specifications, and what @benHeid faced in the other issue, personally I'll feel safer (definitely easier) for leaving this choice to user to configure rather than adding a frequency inference logic in `sktime`.

Comment (fkiraly) [reply sktime/forecasting/base/adapters/_neuralforecast.py:143]: yes, but what I mean, can you not just take it from `fh`? I.e., instea dof the `__init__` in `_fit`, you take `fh.freq`.

What happens, say, if the `freq` in `y` and that of the `__init__` do not agree? Does that not lead to wrong behaviour?

File: sktime/forecasting/neuralforecast.py (hunk 2 above)
Comment (fkiraly) [reply sktime/forecasting/neuralforecast.py:132]: I mean not concrete tests but dependencies. That is, how does the end state of `pyproject.toml` look like?

File: sktime/forecasting/base/adapters/_neuralforecast.py (hunk 1 above)
Comment (fkiraly) [reply sktime/forecasting/base/adapters/_neuralforecast.py:180]: you seem to be skipping the in-sample tests in `tests/_config` - that should be unnecessary if you raise this error.

File: sktime/forecasting/base/adapters/_neuralforecast.py (hunk 3 above)
Comment (jmoralez) [reply sktime/forecasting/base/adapters/_neuralforecast.py:143]: The `freq` is just used to build the future timestamps when calling `predict`. Since you don't use those timestamps you could just set `freq='D'` internally, for example.

File: sktime/forecasting/base/adapters/_neuralforecast.py (hunk 1 above)
Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_neuralforecast.py:180]: I think I tested locally earlier and it failed, but didn't check if that skip is necessary after these lines. Will remove that skip from there.

File: sktime/forecasting/neuralforecast.py (hunk 2 above)
Comment (yarnabrina) [reply sktime/forecasting/neuralforecast.py:132]: I completely misunderstood then.

I am open for suggestions. Here's what I already have though of.

1. I can leave as it is in `forecasting` extra, as I'd guess a lot of people may be interested in using `neuralforecast` models, anything automl (they have AutoRNN, AutoTFT, etc. quite a few) seems to be popular.
2. I can move to `dl`, but that means it will not be covered by tests once we retire old CI. I can add a new job in CI specifically to test `dl` extra, because I guess it needs to be tested not just in `forecasting` but `network` and `regression` as well. Please correct me if that's a wrong idea.

I am of course biased, but I'm tempted to do (1). An excuse is mainly that it'll allow us to have `ray[tune]` as well, which may invite others to contribute a new tuner similar to what @hazrulakmal did for skopt.

In any case, probably I should remove it from `all_extras` and `all_extras_pandas2` at least?

File: sktime/forecasting/base/adapters/_neuralforecast.py (hunk 3 above)
Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_neuralforecast.py:143]: > The freq is just used to build the future timestamps when calling predict. Since you don't use those timestamps you could just set freq='D' internally, for example.

Hi @jmoralez, I am unable to follow how passing a fixed freuqency, e.g. `D` will suffice. I posted an example above (direct use of neuralforecast), where it seems to generate wrong dates if frequency is incorrect. Can you please elaborate?

Comment (jmoralez) [reply sktime/forecasting/base/adapters/_neuralforecast.py:143]: I saw that you only extract the prediction column (`model_point_predictions = model_forecasts[prediction_column_names[0]].to_numpy()`) from the dataframe that neuralforecast returns, so I don't think the timestamps it returns matter, do they?

Those timestamps are there just to make it easier for a user to merge with a validation set or similar, the predictions will be the same if you set different frequencies (only the timestamps will change).

Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_neuralforecast.py:143]: Okay, let me fix it at `D` and trigger all tests.

Comment (jmoralez) [reply sktime/forecasting/base/adapters/_neuralforecast.py:143]: Oh, sorry. We do check the timestamps of the `futr_df` when calling predict, that's my bad. Those timestamps need to match the ones generated with the specified `freq` from `__init__`

File: docs/source/api_reference/forecasting.rst [hunk 4]
```diff
@@ -363,6 +363,16 @@ Transformer (deep learning) based forecasters
...
+RNN (deep learning) based forecasters
+-------------------------------------
+
+.. currentmodule:: sktime.forecasting.neuralforecast
+
+.. autosummary::
+    :toctree: auto_generated/
+    :template: class.rst
+
+    NeuralForecastRNN
```
Comment (yarnabrina) [inline docs/source/api_reference/forecasting.rst:375]: @benHeid Please let me know if you have suggestion for this grouping.

(and of course the general review of the PR, especially your opinion on pyproject.toml final state.)

File: sktime/forecasting/base/adapters/_neuralforecast.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+            raise ValueError("Missing exogeneous data, 'futr_exog_list' is non-empty.")
+
+        if self.futr_exog_list:
+            for column in self.futr_exog_list:
+                train_data[column] = X[column].to_numpy()
+
+        train_dataset = pandas.DataFrame(data=train_data)
+
+        maximum_forecast_horizon = fh.to_relative(self.cutoff)[-1]
+        self._forecaster = self._instantiate_model(maximum_forecast_horizon)
```
Comment (jmoralez) [inline sktime/forecasting/base/adapters/_neuralforecast.py:202]: Just a small question, should this be `self.forecaster_`, given that the docstrings for fit and predict say *Fitted model attributes ending in "_"*?

Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_neuralforecast.py:202]: As of now, all other adapters (statsmodels/pmdarima/etc.) use `self._forecaster`. I've not changed here or in the `statsforecast` adapter to remain consistent.

I think the reason for not doing in adapters is that having underscore at the end updates the results of `get_fitted_params()`, and fitted model itself is not a parameter. This is my guess, may be @fkiraly / @benHeid can confirm if this was the reason for earlier adapters or something else.

But having the model itself, or if it has some other properties like AIC/BIC/etc., may be useful. I'll wait for @fkiraly / @benHeid to comment for fitted model, but if you can suggest some other quick model properties that may be useful for users, I can override `_get_fitted_params` in the meanwhile.

File: sktime/forecasting/base/adapters/_neuralforecast.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        "X-y-must-have-same-index": True,
+        "handles-missing-data": False,
+        "capability:insample": False,
+    }
+
+    def __init__(
+        self: "_NeuralForecastAdapter",
+        freq: str,
+        local_scaler_type: typing.Optional[
+            typing.Literal["standard", "robust", "robust-iqr", "minmax"]
```
Comment (jmoralez) [inline sktime/forecasting/base/adapters/_neuralforecast.py:71]:
```suggestion
            typing.Literal["standard", "robust", "robust-iqr", "minmax", "boxcox"]
```

File: sktime/forecasting/neuralforecast.py [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+        # estimator type
+        # --------------
+        "python_dependencies": ["neuralforecast>=1.6.4"],
+    }
+
+    def __init__(
+        self: "NeuralForecastRNN",
+        freq: str,
+        local_scaler_type: typing.Optional[
+            typing.Literal["standard", "robust", "robust-iqr", "minmax"]
```
Comment (jmoralez) [inline sktime/forecasting/neuralforecast.py:161]:
```suggestion
            typing.Literal["standard", "robust", "robust-iqr", "minmax", "boxcox"]            
```

**Review by jmoralez — APPROVED** [review]
LGTM on the nixtla side. Thanks for working on this!

File: sktime/forecasting/base/adapters/_neuralforecast.py (hunk 6 above)
Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_neuralforecast.py]: Thanks, added. It's not showing here for some reason.

https://nixtlaverse.nixtla.io/neuralforecast/core.html#neuralforecast

Comment (jmoralez) [reply sktime/forecasting/base/adapters/_neuralforecast.py]: We removed it in a previous PR but [this one](https://github.com/Nixtla/neuralforecast/pull/896) adds it back.

File: sktime/forecasting/base/adapters/_neuralforecast.py (hunk 5 above)
Comment (fkiraly) [reply sktime/forecasting/base/adapters/_neuralforecast.py:202]: really depends on what you would like to expose to the user. It is a valid choice to not expose the adapted class, as long as most fitted paraemters are forwarded, but I think it would also be valid to do so.

As @yarnabrina says, in the current codebase, components (parameters) usually have their fitted clones exposed as a fitted parameters, but adaptees not, although I don't think there's a strict rule on this.

**Review by fkiraly — APPROVED** [review]
Looks good, thx!
